### ISSUE TITLE
length of masters array always bigger than 0?

### ISSUE THREAD

**Issue by crucidal** [mesosphere/mesos-dns#276]
When using a zk string instead of a masters string the console gets spammed with the following message:
ERROR: 2015/09/23 10:55:03 generator.go:342: warning: leader "master@xxx.xxx.xxx.xxx:yyyy" is not in master list

omitting the masters variable or creating an empty array ("masters": []) generates this error/warning.

**Comment by crucidal** [comment]
after some testing I noticed something obvious which I should have seen before...
I've defined DNS names in my zk string and master string.
However, the error message complains that some ip adress is not in the masters string.
I'd expect there to be no error message when I omit the masters string but at least it makes sense when I do define one with dns names.

Can this be fixed? :-)

**Comment by s-urbaniak** [comment]
I was able to reproduce the issue with a local 3-master setup. 

When parsing the current leader from `state.json` during the periodic reload we get:
```
$ curl -s 127.0.0.1:5050/master/state.json | jq '.leader'
"master@127.0.0.1:5050"
```

This doesn't quite match what zookeeper gives us:
```
I0924 16:07:03.726162    2763 detect.go:219] notifying of master membership change: [
  &MasterInfo{
    Id:*20150924-150935-50331775-5050-1411,
    Ip:*50331775,Port:*5050,
    Pid:*master@127.0.0.3:5050,
    Hostname:*127.0.0.3,
    Version:*0.23.0,XXX_unrecognized:[],}
  &MasterInfo{
    Id:*20150924-150932-33554559-5050-1410,
    Ip:*33554559,Port:*5050,
    Pid:*master@127.0.0.2:5050,
    Hostname:*127.0.0.2,
    Version:*0.23.0,XXX_unrecognized:[],}
  &MasterInfo{
    Id:*20150924-150927-16777343-5050-1402,
    Ip:*16777343,
    Port:*5050,
    Pid:*master@127.0.0.1:5050,
    Hostname:*localhost,
    Version:*0.23.0,XXX_unrecognized:[],}
]
```

There is an apparent string mismatch between `"localhost"` (what zookeper gives us) and `"127.0.0.1"` (what `state.json` gives us) causing the observed warning because "127.0.0.1" is not in the master list ["localhost", "127.0.0.2", "127.0.0.3"].

**Comment by jdef** [comment]
why do you have one master registering with --hostname=localhost and the
others with an IP address? or is that not the case?

**Comment by s-urbaniak** [comment]
I'm not using `--hostname` at all, I use `--ip` only in my local playground:

```
$ /usr/local/sbin/mesos-master \
  --registry=in_memory \
  --quorum=2 \
  --ip=127.0.0.[1,2,3] \
  --zk='zk://127.0.0.1:2181/mesos'
```

(having 127.0.0.2 and 127.0.0.3 defined as aliases)

@crucidal can you post your mesos-master configuration, especially the `--hostname` and `--ip` settings?

**Comment by crucidal** [comment]
We're working with 'cloud machines' so instead of using (static) IP's we're using dns names for everything.
so `--hostname` and the entries in `--zk` are all dns names. we're not using `--ip`.

```
/usr/sbin/mesos-master --zk=zk://clrv004593.example.com:2181,clrv004595.example.com:2181,clrv004597.example.com:2181/mesos 
--quorum=2 
--work_dir=/var/lib/mesos 
--port=5050 
--log_dir=/var/log/mesos 
--hostname=clrv004593.example.com
```

**Comment by s-urbaniak** [comment]
@jdef might that be the root cause? `--hostname` propagating those hostnames ("clrv004593.example.com") to zookeeper and `/master/state.json` reporting back IPs?

**Comment by jdef** [comment]
perhaps. I wonder what the state.json's Leader is reporting

**Comment by s-urbaniak** [comment]
@crucidal Do you mind to check if `curl -s [MASTER_IP]:5050/master/state.json | jq '.leader'` returns an IP address?

**Comment by sepiroth887** [comment]
There is a chance your env has hostname set to localhost as usually 127.0.0.1 corresponds to that on some systems by default ? 

Setting hostname when using a loopback might help

**Comment by s-urbaniak** [comment]
@sepiroth887 yes indeed that's the case, localhost resolves to 127.0.0.1.

I did some more debugging and either I'm doing something wrong or we have some strange invariants in mesos-dns.

mesos-dns's internal state with the following setup is a colorful mixture of IPs and hostnames causing the above log statement to be generated.

When I start three mesos masters with a given `--ip` only, leaving out `--hostname` (which according to the mesos documentation is valid, quote "If left unset, the hostname is resolved from the IP address that the slave binds to")
```
$ /usr/local/sbin/mesos-master --ip=127.0.0.2 --zk='zk://127.0.0.1:2181/mesos' --quorum=2 --registry=in_memory
$ /usr/local/sbin/mesos-master --ip=127.0.0.3 --zk='zk://127.0.0.1:2181/mesos' --quorum=2 --registry=in_memory
$ /usr/local/sbin/mesos-master --ip=127.0.0.4 --zk='zk://127.0.0.1:2181/mesos' --quorum=2 --registry=in_memory
```
Having the following entries in `/etc/hosts`:
```
127.0.0.2   localhost2
127.0.0.3   localhost3
127.0.0.4   localhost4
```
And mesos giving me the following leader:
```
$ curl 127.0.0.2:5050/master/state.json | jq .leader
"master@127.0.0.2:5050"
```
When I start mesos-dns the following master list is retrieved via zookeeper:
```
masters.go:61: Updated masters:  [
  &MasterInfo{
    Id:*20150926-175458-33554559-5050-62952,
    Ip:*33554559,
    Port:*5050,
    Pid:*master@127.0.0.2:5050,
    Hostname:*localhost2,
    Version:*0.24.0,
    XXX_unrecognized:[],
  }
  &MasterInfo{
    Id:*20150926-175459-50331775-5050-62953,
    Ip:*50331775,
    Port:*5050,
    Pid:*master@127.0.0.3:5050,
    Hostname:*localhost3,
    Version:*0.24.0,
    XXX_unrecognized:[],
  }
  &MasterInfo{
    Id:*20150926-175500-67108991-5050-62954,
    Ip:*67108991,
    Port:*5050,
    Pid:*master@127.0.0.4:5050,
    Hostname:*localhost4,
    Version:*0.24.0,
    XXX_unrecognized:[],
  }
]
```
I end up with the following bogus `records.RecordGenerator` state:
```
&{
  As:map[
    master3.mesos.:[127.0.0.2] 
    ns1.mesos.:[127.0.0.1] 
    leader.mesos.:[127.0.0.2] 
    master.mesos.:[127.0.0.2 localhost2 localhost4 localhost3] 
    master0.mesos.:[localhost2] 
    master1.mesos.:[localhost4] 
    master2.mesos.:[localhost3]
  ]
  SRVs:map[
    _leader._tcp.mesos.:[leader.mesos.:5050] 
    _leader._udp.mesos.:[leader.mesos.:5050]
  ] 
  SlaveIPs:map[]
}
```
When I change https://github.com/mesosphere/mesos-dns/blob/master/detect/masters.go#L104 to prefer IPs over hostnames everything works correctly.

PS: https://github.com/mesosphere/mesos-dns/blob/master/detect/masters.go#L109 BigEndian must be changed to LittleEndian.

I'll open a separate PR with "prefer IP changes" and a separate PR for the endianness fix.

**Comment by jdef** [comment]
xref https://issues.apache.org/jira/browse/MESOS-1201

**Comment by crucidal** [comment]
@s-urbaniak `curl -s localhost:5050/master/state.json | jq '.leader'` returns master@{IP}:{PORT}
it makes no difference if I use `localhost` or when I use another machine and route the call through nginx using a `domain name`.

seems like you're working hard on solving the issue :-) keep it up :+1:

**Comment by s-urbaniak** [comment]
@crucidal Do you mind rechecking with the current master version of mesos-dns?

1. Those log entries should be gone
2. `dig @127.0.0.1 -p 5253 master.mesos +short A` should give you a list of all master IP entries
3. `dig @127.0.0.1 -p 5253 leader.mesos +short A` should give you the leader IP only.

PS: your `master@{IP}:{PORT}` response doesn't look right.

**Comment by s-urbaniak** [comment]
@crucidal ping :-)

**Comment by crucidal** [comment]
@s-urbaniak 
Hi,
sorry for taking so much time to respond!
I haven't got round to testing this with the new version yet as we've been implementing an alternative solution for the time being. Since it's so easy to implement and test your solution I'll try to see if I can do that somewhere in the next 2 sprints (2 - 4 weeks).

in case your interested why we didn't choose mesos-dns for now:
1. I couldn't figure out how to configure nginx to use mesos-dns for domain x and another name server for domain y. (to enable both containers in the cluster as well as users outside of it to use the dns names)
2. getting the service port still required a script of some sort.
3. the errors aren't catastrophic and I'm sure new ones will be fixed just as quickly as this one. but I wasn't comfortable enough to put it in production just yet.

for now we use a small javascript program which co-exists in a docker container next to Nginx. to update a conf folder and the information comes from the marathon even stream.

So, once again. 2 - 4 weeks. and keep this project going! hopefully we can use it in the future :-)

Kind Regards,
Crucidal

**Comment by sargun** [comment]
@crucidal Bump.

**Comment by crucidal** [comment]
It's been a while...
This issue has been resolved and works properly in v0.5.1
thanks.
